src/transformers/models/gpt_neo/modeling_gpt_neo.py

```diff
@@ -0,0 +1,946 @@
+        else:
+            present = None
+
+        attn_outputs = self._attn(query, key, value, attention_mask, head_mask, output_attentions)
+        a = attn_outputs[0]
+
+        a = self.merge_heads(a)
+        a = self.out_proj(a)
+        a = self.resid_dropout(a)
+
+        return (a, present) + attn_outputs[1:]  # a, present, (attentions)
+
+
+class LocalAttention(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+
+        self.register_buffer("masked_bias", torch.tensor(-1e9))
+
+        self.attn_dropout = nn.Dropout(config.attention_dropout)
+        self.resid_dropout = nn.Dropout(config.resid_dropout)
+
+        self.embed_dim = config.hidden_size
+        self.num_heads = config.num_heads
+        self.head_dim = self.embed_dim // self.num_heads
+        assert (
+            self.head_dim * self.num_heads == self.embed_dim
+        ), f"embed_dim must be divisible by num_heads (got `embed_dim`: {self.embed_dim} and `num_heads`: {self.num_heads})."
+
+        self.k_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=False)
+        self.v_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=False)
+        self.q_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=False)
+        self.out_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=True)
+
+        self.window_size = config.window_size
+
+    def shift(self, x, offset, pad_value=0, dim=2):
+        t = x.shape[1]
+        dims = (len(x.shape) - dim) * (0, 0)
+        padded_x = F.pad(x, (*dims, offset, 0), value=pad_value)
+        return padded_x[:, :t, ...]
+
+    def look_around(self, x, block_length, window_size):
+        num_complete_blocks = window_size // block_length
+
+        parts = [x]
+        for i in range(1, num_complete_blocks + 1):
+            parts = [self.shift(x, i)] + parts
+
+        partial_size = window_size % block_length
+        if partial_size > 0:
+            margin = x[:, :, block_length - partial_size : block_length, ...]
+            parts = [self.shift(margin, num_complete_blocks + 1)] + parts
+        return torch.cat(parts, dim=2)
+
+    def split_heads(self, x, k=False):
+        new_x_shape = x.size()[:-1] + (self.num_heads, x.size(-1) // self.num_heads)
+        x = x.view(*new_x_shape)
+        if k:
+            return x.permute(0, 1, 3, 4, 2)  # (batch, chunks, head, head_features, seq_length)
+        else:
+            return x.permute(0, 1, 3, 2, 4)  # (batch, chunks, head, seq_length, head_features)
+
+    def merge_heads(self, x):
+        x = x.permute(0, 1, 3, 2, 4).contiguous()
+        new_x_shape = x.size()[:-2] + (x.size(-2) * x.size(-1),)
+        return x.view(*new_x_shape)
+
+    def create_buckets(self, tensor, num_buckets, block_length):
+        return tensor.reshape(tensor.size()[0], num_buckets, block_length, -1)
+
+    def create_attention_mask(self, bs, seq_len, windows, block_length, attention_mask):
+        ticker = torch.arange(seq_len)[None, :]
```

Comment:
very hard to understand this function...a refactor would be useful here with better namings and more comments